src/SelfServe/SelfServeUtils.tsx

```diff
@@ -0,0 +1,210 @@
+import {
+  ChoiceItem,
+  Node,
+  Info,
+  InputTypeValue,
+  Descriptor,
+  AnyInput,
+  NumberInput,
+  StringInput,
+  BooleanInput,
+  ChoiceInput,
+  InputType
+} from "../Explorer/Controls/SmartUi/SmartUiComponent";
+
+const SelfServeType = "selfServeType";
+
+export class SelfServeBase {
+  public static toSmartUiDescriptor(): Descriptor {
+    return Reflect.getMetadata(this.name, this) as Descriptor;
+  }
+}
+
+export interface CommonInputTypes {
+  id: string;
+  info?: (() => Promise<Info>) | Info;
+  parentOf?: string[];
+  type?: InputTypeValue;
+  label?: (() => Promise<string>) | string;
+  placeholder?: (() => Promise<string>) | string;
+  dataFieldName?: string;
+  min?: (() => Promise<number>) | number;
+  max?: (() => Promise<number>) | number;
+  step?: (() => Promise<number>) | number;
+  trueLabel?: (() => Promise<string>) | string;
+  falseLabel?: (() => Promise<string>) | string;
+  choices?: (() => Promise<ChoiceItem[]>) | ChoiceItem[];
+  inputType?: string;
+  onChange?: (currentState: Map<string, InputType>, newValue: InputType) => Map<string, InputType>;
+  onSubmit?: (currentValues: Map<string, InputType>) => Promise<void>;
+  initialize?: () => Promise<Map<string, InputType>>;
+  customElement?: ((currentValues: Map<string, InputType>) => Promise<JSX.Element>) | JSX.Element;
+}
+
+const setValue = <T extends keyof CommonInputTypes, K extends CommonInputTypes[T]>(
+  name: T,
+  value: K,
+  fieldObject: CommonInputTypes
+): void => {
+  fieldObject[name] = value;
+};
+
+const getValue = <T extends keyof CommonInputTypes>(name: T, fieldObject: CommonInputTypes): unknown => {
+  return fieldObject[name];
+};
+
+export const addPropertyToMap = (
+  target: Object,
+  propertyKey: string,
+  metadataKey: string,
+  descriptorName: string,
+  descriptorValue: any
+): void => {
+  const descriptorKey = descriptorName.toString() as keyof CommonInputTypes;
+  let context = Reflect.getMetadata(metadataKey, target) as Map<string, CommonInputTypes>;
+
+  if (!context) {
+    context = new Map<string, CommonInputTypes>();
+  }
+
+  if (!(context instanceof Map)) {
+    throw new Error("@SmartUi should be the first decorator for the class.");
+  }
+
+  let propertyObject = context.get(propertyKey);
+  if (!propertyObject) {
+    propertyObject = { id: propertyKey };
+  }
+
+  if (getValue(descriptorKey, propertyObject) && descriptorKey !== "type" && descriptorKey !== "dataFieldName") {
+    throw new Error("duplicate descriptor");
+  }
+
+  setValue(descriptorKey, descriptorValue, propertyObject);
+  context.set(propertyKey, propertyObject);
+
+  Reflect.defineMetadata(metadataKey, context, target);
+};
+
+export const toSmartUiDescriptor = (metadataKey: string, target: Object): void => {
+  const context = Reflect.getMetadata(metadataKey, target) as Map<string, CommonInputTypes>;
+  Reflect.defineMetadata(metadataKey, context, target);
+
+  const root = context.get("root");
+  context.delete("root");
```

Comment:
this will remove the intermediate data structure created for this particular self serve class (a map built using the descriptors). This needs to be removed so that the class can be recompiled and the structure rebuilt on the next load.